src/operations/search/search.js

```diff
@@ -207,108 +208,162 @@ module.exports.search = async (requestInfo, args, resourceName, collection_name)
-            // https://www.hl7.org/fhir/search.html#total
-            // if _total is passed then calculate the total count for matching records also
-            // don't use the options since they set a limit and skip
-            if (args['_total'] === 'estimate') {
-                total_count = await collection.estimatedDocumentCount(query, {maxTimeMS: maxMongoTimeMS});
-            } else {
-                total_count = await collection.countDocuments(query, {maxTimeMS: maxMongoTimeMS});
-            }
-        }
-        // Resource is a resource cursor, pull documents out before resolving
         /**
          * resources to return
          * @type {Resource[]}
          */
         let resources = [];
-        while (await cursor.hasNext()) {
+        /**
+         * @type {number}
+         */
+        let total_count = 0;
+        /**
+         * which index hint to use (if any)
+         * @type {string|null}
+         */
+        let indexHint = null;
+        /**
+         * @type {int | null}
+         */
+        let cursorBatchSize = null;
+        // run the query and get the results
+        if (query !== null) {
+            // Now run the query to get a cursor we will enumerate next
+            let cursorQuery = await collection.find(query, options).maxTimeMS(maxMongoTimeMS);
+
+            // set batch size if specified
+            if (env.MONGO_BATCH_SIZE || args['_cursorBatchSize']) {
+                cursorBatchSize = args['_cursorBatchSize'] ? parseInt(args['_cursorBatchSize']) : parseInt(env.MONGO_BATCH_SIZE);
+                if (cursorBatchSize > 0) {
+                    cursorQuery = cursorQuery.batchSize(cursorBatchSize);
+                }
+            }
             /**
-             * element
-             * @type {Resource}
+             * mongo db cursor
+             * @type {Promise<Cursor<unknown>> | *}
              */
-            const element = await cursor.next();
-            if (args['_elements']) {
-                /**
-                 * @type {string}
-                 */
-                const properties_to_return_as_csv = args['_elements'];
-                /**
-                 * @type {string[]}
-                 */
-                const properties_to_return_list = properties_to_return_as_csv.split(',');
+            let cursor = await pRetry(
+                async () =>
+                    await cursorQuery,
+                {
+                    retries: 5,
+                    onFailedAttempt: async error => {
+                        let msg = `Search ${resourceName}/${JSON.stringify(args)} Retry Number: ${error.attemptNumber}: ${error.message}`;
+                        logError(user, msg);
+                        await logMessageToSlack(msg);
+                    }
+                }
+            );
+
+            // find columns being queried and match them to an index
+            if (isTrue(env.SET_INDEX_HINTS) || args['_setIndexHint']) {
```

Comment:
In a future state, should we create modules/functions for each of the 'if' statements in this method? This way it makes the code smaller readable/testable modules.